### PR TITLE
Add new FL sources to MCD/Compound

### PR DESCRIPTION
**Pull request by yippee-ki-yay**
Remove old AaveV1 FL and add AaveV2/Balancer with the options for the backend to choose which one the they want to use.
The change was aimed to be minimal and to only add the new FL, the biggest difference is that the old system was getting a FL in ETH but here we get it in WETH.

A few notes:
- Tests are working but are ran from a hard coded block number written in test comments
- Added a separate FLHelper file for common stuff needed between files (view for backend there as well)

### CODE REVIEW

File: contracts/mcd/saver/MCDSaverProxy.sol [hunk 1]
```diff
@@ -190,7 +190,8 @@ contract MCDSaverProxy is DFSExchangeCore, MCDSaverProxyHelper {
         address _managerAddr,
         uint256 _cdpId,
         address _joinAddr,
-        uint256 _amount
+        uint256 _amount,
+        bool _toEth
```
Comment (majkic99) [inline contracts/mcd/saver/MCDSaverProxy.sol:194]: add param to natspec

File: contracts/compound/automatic/CompoundMonitor.sol [hunk 2]
```diff
@@ -30,6 +32,7 @@ contract CompoundMonitor is AdminAuth, DSMath, CompoundSafetyRatio {
     CompoundMonitorProxy public compoundMonitorProxy = CompoundMonitorProxy(0xB1cF8DE8e791E4Ed1Bd86c03E2fc1f14389Cb10a);
     CompoundSubscriptions public subscriptionsContract = CompoundSubscriptions(0x52015EFFD577E08f498a0CCc11905925D58D6207);
     address public compoundFlashLoanTakerAddress;
+    address internal compoundSaverFlashLoan;
```
Comment (majkic99) [inline contracts/compound/automatic/CompoundMonitor.sol:35]: should one be public one internal?

**Review by majkic99 — APPROVED** [review]
test works, lgtm, small changes maybe needed

File: contracts/compound/automatic/CompoundMonitor.sol (hunk 2 above)
Comment (yippee-ki-yay) [reply contracts/compound/automatic/CompoundMonitor.sol:35]: I set it internal for deployment gas cost, the old one is public and would keep it that way for easier finding on etherscan when looking which taker addr is called and the second contract can be found in the taker.

**Review by kjaredic — APPROVED** [review]
legitimate

File: contracts/compound/saver/CompoundSaverFlashLoan.sol [hunk 3]
```diff
@@ -1,32 +1,65 @@
...
+
+    // AaveV2 fl callback
+     function executeOperation(
+        address[] memory _assets,
+        uint256[] memory _amounts,
+        uint256[] memory _fees,
+        address,
+        bytes memory _params
+    ) public returns (bool) {
+        require(msg.sender == AAVE_LENDING_POOL, "Only aave pool can call");
```
Comment (nklipa13) [inline contracts/compound/saver/CompoundSaverFlashLoan.sol:43]: Is it better to use:
`ILendingPoolAddressesProviderV2(_market).getLendingPool()` ?

File: contracts/mcd/saver/MCDSaverFlashLoan.sol [hunk 4]
```diff
@@ -21,12 +19,44 @@ contract MCDSaverFlashLoan is MCDSaverProxy, AdminAuth, FlashLoanReceiverBase {
...
+
+    // AaveV2 fl callback
+     function executeOperation(
+        address[] memory _assets,
+        uint256[] memory _amounts,
+        uint256[] memory _fees,
+        address,
+        bytes memory _params
+    ) public returns (bool) {
+        require(msg.sender == AAVE_LENDING_POOL, "Only aave pool can call");
```
Comment (nklipa13) [inline contracts/mcd/saver/MCDSaverFlashLoan.sol:44]: The same question about
`ILendingPoolAddressesProviderV2(_market).getLendingPool()`

File: contracts/mcd/saver/MCDSaverTaker.sol [hunk 5]
```diff
@@ -37,29 +34,42 @@ contract MCDSaverTaker is MCDSaverProxy {
...
-        MCD_SAVER_FLASH_LOAN.transfer(msg.value); // 0x fee
-
-        Manager(managerAddr).cdpAllow(_cdpId, MCD_SAVER_FLASH_LOAN, 1);
-
-        bytes memory paramsData = abi.encode(packExchangeData(_exchangeData), _cdpId, _gasCost, _joinAddr, false, uint8(_managerType));
-
-        lendingPool.flashLoan(MCD_SAVER_FLASH_LOAN, DAI_ADDRESS, loanAmount, paramsData);
-
-        Manager(managerAddr).cdpAllow(_cdpId, MCD_SAVER_FLASH_LOAN, 0);
+        _mcdSaverFlashLoanAddr.transfer(msg.value); // 0x fee
```
Comment (nklipa13) [inline contracts/mcd/saver/MCDSaverTaker.sol:37]: We can remove this

File: contracts/mcd/saver/MCDSaverTaker.sol [hunk 6]
```diff
@@ -73,23 +83,31 @@ contract MCDSaverTaker is MCDSaverProxy {
         uint256 loanAmount = sub(_exchangeData.srcAmount, maxColl);
         loanAmount = loanAmount > maxLiq ? maxLiq : loanAmount;
 
-        MCD_SAVER_FLASH_LOAN.transfer(msg.value); // 0x fee
-
-        Manager(managerAddr).cdpAllow(_cdpId, MCD_SAVER_FLASH_LOAN, 1);
-
-        bytes memory paramsData = abi.encode(packExchangeData(_exchangeData), _cdpId, _gasCost, _joinAddr, true, uint8(_managerType));
+        _mcdSaverFlashLoanAddr.transfer(msg.value); // 0x fee
```
Comment (nklipa13) [inline contracts/mcd/saver/MCDSaverTaker.sol:86]: We can also remove this

File: contracts/utils/FLHelper.sol [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        BALANCER,
+        NO_LOAN
+    }
+
+    function _getFL(
+        FLType _flType,
+        address _tokenAddr,
+        uint256 _flAmount,
+        bytes memory _callData,
+        address MCD_SAVER_FLASH_LOAN
```
Comment (nklipa13) [inline contracts/utils/FLHelper.sol:29]: Rename this param

File: contracts/utils/FLHelper.sol [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            IFlashLoans(BALANCER_VAULT_ADDR).flashLoan(MCD_SAVER_FLASH_LOAN, tokens, amounts, _callData);
+        }
+    }
+
+    function getProtocolLiq(address _tokenAddr, uint256 _desiredAmount) public view returns (FLType flType) {
+        uint256 flLiquidity = getFLLiquidity(FLType.BALANCER, _tokenAddr);
+
+        if (flLiquidity >= _desiredAmount) return FLType.BALANCER;
+
+        flLiquidity = getFLLiquidity(FLType.BALANCER, _tokenAddr);
```
Comment (nklipa13) [inline contracts/utils/FLHelper.sol:65]: Here it should be FLType.AAVE_V2

**Review by nklipa13 — CHANGES_REQUESTED** [review]
Mostly small improvements and one fix in FLHelper needed before merge

**Review by nklipa13 — APPROVED** [review]

**Pull request merged**